parquet source with partition extraction

unical1988 opened the pull request:
Added ConversionSource for Parquet files
- Partitions format are specified by a user config.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,193 @@
+    public List<InternalDataFile> getInternalDataFiles() {
+        List<InternalDataFile> internalDataFiles = null;
+        List<LocatedFileStatus> parquetFiles = getParquetFiles(hadoopConf, basePath);
+        InternalTable table = getTable(-1L);
```
Remove this unused line

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,193 @@
+                .build();
+    }
+
+    public List<InternalDataFile> getInternalDataFiles() {
```
Let's make this private

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,193 @@
+    @Override
+    public InternalSnapshot getCurrentSnapshot() {
+        List<InternalDataFile> internalDataFiles = getInternalDataFiles();
+        InternalTable table = getTable(-1L);
```
Since this call also requires listing the data files, let's make a new private method for `getTable` that takes in the list of files so it doesn't get read twice

unical1988 replied:
I refactored as pointed out but wouldn't make InternalTable extractor getTable()) not usable

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,193 @@
+    public InternalSnapshot getCurrentSnapshot() {
+        List<InternalDataFile> internalDataFiles = getInternalDataFiles();
+        InternalTable table = getTable(-1L);
+        return InternalSnapshot.builder()
```
We'll need to set the `version` here. I am guessing it should be the last modification time but I need to think it through more.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,193 @@
+    }
+
+
+    public List<LocatedFileStatus> getParquetFiles(Configuration hadoopConf, String basePath) {
```
Let's make this private as well

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,193 @@
+                                                                        hadoopConf, file.getPath())))
+                                                .build())
+                        .collect(Collectors.toList());
+        return internalDataFiles;
```
Nitpick: just return instead of assigning the output to a local variable

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,193 @@
+
+    // since we are considering files instead of tables in parquet
+    @Override
+    public CommitsBacklog<java.lang.Long> getCommitsBacklog(
```
For all usages of `java.lang.Long` you can shorten it to `Long`

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,193 @@
+
+@Builder
+// @NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class ParquetConversionSource implements ConversionSource<Long> {
```
For us to provide support for incremental sync, I think this may need to be a range instead of a singular point. That way we can filter the files by a start and end time when performing the sync.

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,256 @@
+            .collect(Collectors.toList());
+    InternalTable internalTable = getMostRecentTable(parquetFiles);
+    List<FileStatus> keptFiles = new ArrayList<>();
+    List<FileStatus> removedFiles = new ArrayList<>(tableChangesBefore);
```
They might potentially be removed, please check `removedFiles.removeAll(keptFiles);` where the `keptFiles `are saved.

the-other-tim-brown replied:
The candidates are all from the call to list the directory though which means they are present.

unical1988 replied:
To identify a file as removed can I consider this logic: "if a file wasn't found in tableChangesAfter and it was in the list of tableChangesBefore then  it was removed? those are as follows:


```
 List<FileStatus> tableChangesAfter =
        parquetFiles.stream()
            .filter(fileStatus -> fileStatus.getModificationTime() > modificationTime)
            .collect(Collectors.toList());
    List<FileStatus> tableChangesBefore =
        parquetFiles.stream()
            .filter(fileStatus -> fileStatus.getModificationTime() < modificationTime)
            .collect(Collectors.toList());
```

the-other-tim-brown replied:
The files are required to be immutable for this to work so you would remove all the files from the previous sync on every run which means you are losing the data.

unical1988 replied:
Ok, so filesDiff should account only for added Files and keep filesRemoved() empty?

the-other-tim-brown replied:
yes

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,256 @@
+
+  @Override
+  public String getCommitIdentifier(Long aLong) {
+    return null;
```
what `getCommitIdentifier()` should do?

the-other-tim-brown replied:
it returns a string that identifies the commit. In Iceberg this is a snapshot ID, in Hudi it is an "Instant" which is just a point in time. In this case there is concept of a commit so we can just use the long that represents the time the sync is run.

the-other-tim-brown replied:
Yes that should work

unical1988 commented on xtable-api/src/main/java/org/apache/xtable/model/schema/PartitionFieldSpec.java:
```diff
@@ -0,0 +1,28 @@
+import lombok.Value;
+
+@Value
+public class PartitionFieldSpec {
```
Sure

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetPartitionValueExtractor.java:
```diff
@@ -0,0 +1,232 @@
+
+
+
+  public List<PartitionValue> extractPartitionValues(
```
there are not exactly the same, I did refactor the class

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSourcePartitionSpecExtractor.java:
```diff
@@ -0,0 +1,31 @@
+ * Partition spec extractor interface specifically designed for Parquet to parse partition values
+ * appropriately.
+ */
+public interface ParquetSourcePartitionSpecExtractor extends SourcePartitionSpecExtractor {
```
done

unical1988 commented on xtable-api/src/main/java/org/apache/xtable/model/schema/PartitionFieldSpec.java:
```diff
@@ -0,0 +1,28 @@
+import lombok.Value;
+
+@Value
+public class PartitionFieldSpec {
```
Ok

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetPartitionSpecExtractor.java:
```diff
@@ -0,0 +1,76 @@
+// @AllArgsConstructor
+public class ParquetPartitionSpecExtractor implements PathBasedPartitionSpecExtractor {
+  private static final ParquetPartitionSpecExtractor INSTANCE = new ParquetPartitionSpecExtractor();
+  private final ParquetSourceConfig config = null; // TODO shouldn't be init to null?
```
Yes

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/ConfigurationBasedPartitionSpecExtractor.java:
```diff
@@ -29,20 +29,21 @@
 @AllArgsConstructor
-public class ConfigurationBasedPartitionSpecExtractor implements HudiSourcePartitionSpecExtractor {
+public class ConfigurationBasedPartitionSpecExtractor implements PathBasedPartitionSpecExtractor {
   private final HudiSourceConfig config;
```
Can we update this to take in the `List<PartitionFieldSpec>` directly so we can reuse it for the parquet source as well?

the-other-tim-brown replied:
Thanks! That is what I was looking for

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,264 @@
+        ParquetMetadata parquetMetadata =
+                parquetMetadataExtractor.readParquetMetadata(hadoopConf, latestFile.getPath());
+
+        List<InternalPartitionField> partitionFields =
```
The `ParquetPartitionSpecExtractor` should be used here. The spec extractor is used for defining which fields are used and the `ParquetPartitionValueExtractor` should be used to get the values for the partition fields per file.

unical1988 replied:
in this case `partitionValueExtractor.extractParquetPartitions()` is useless?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,264 @@
+        InternalTable internalTable = getMostRecentTable(parquetFiles);
+
+        for (FileStatus tableStatus : tableChangesAfter) {
+            boolean isPresent = tableChangesBefore.contains(tableStatus);
```
I think we can skip this step as the sets of FileStatus should be distinct due to the condition on the modification time

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/ITConversionController.java:
```diff
@@ -141,7 +140,7 @@ private static Stream<Arguments> testCasesWithPartitioningAndSyncModes() {
   private static Stream<Arguments> generateTestParametersForFormatsSyncModesAndPartitioning() {
     List<Arguments> arguments = new ArrayList<>();
-    for (String sourceTableFormat : Arrays.asList(HUDI, DELTA, ICEBERG)) {
+    for (String sourceTableFormat : Arrays.asList(HUDI, DELTA, ICEBERG, PARQUET)) {
```
This will likely cause some issues since we don't have a concept for updates or deletes in the parquet source. It is assumed to be append-only. We can instead setup a new test in this class that creates parquet tables, converts them to all 3 other table formats, and validates the data using the existing helpers.

unical1988 replied:
I notice you start the tests by creating a SparkSession, can we leverage it for creating the parquet file to convert into 3 other formats? is a call to  `conversionController.sync(conversionConfig, conversionSourceProvider); `sufficient to test the coversion? do you validate the converted data through `checkDatasetEquivalence()` (all of `ITConversionController.java`) ?

the-other-tim-brown replied:
Yes we can use the spark session and your understanding of the methods is correct

unical1988 replied:
Is this the example to follow for testing? https://github.com/apache/incubator-xtable/blob/19cbda42cca1504fcfbaf130dabce8b1f7b24399/xtable-core/src/test/java/org/apache/xtable/ITConversionController.java#L537

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,255 @@
+        .collect(Collectors.toList());
+  }
+
+  private List<LocatedFileStatus> getParquetFiles(Configuration hadoopConf, String basePath) {
```
ok

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,255 @@
+  }
+
+  @Override
+  public CommitsBacklog<Long> getCommitsBacklog(InstantsForIncrementalSync syncInstants) {
+    // based on either table formats?
+    List<Long> commitsToProcess =
+        Collections.singletonList(syncInstants.getLastSyncInstant().toEpochMilli());
+    return CommitsBacklog.<Long>builder().commitsToProcess(commitsToProcess).build();
+  }
```
is `getParquetFilesInRange() ` not useful here? how to get the files based on a time range? based on which attribute? modificationTime or other?

unical1988 replied:
@the-other-tim-brown for now we limit the source to snapshot syncs

unical1988 commented:
@the-other-tim-brown the test class should create the files using a spark session ?

unical1988 commented:
@the-other-tim-brown could you check with me why the syncs are failing for all 3 formats?

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetParitionUtil.java:
```diff
@@ -0,0 +1,110 @@
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class ParquetParitionUtil {
+
+    public static SparkSession getSparkSession(Path tempDir) {
```
Was also what I was thinking, but for the tests to pass now lets stick to this solution. Also could you check and tell me what writingStatuses in Hudi means? There is an error when syncing involving that

unical1988 replied:
No it is not that, the fail does not relate to the sparksession nor to the sourcefield

unical1988 replied:
The partition field is not in the list of schema.getFields because when you partition a file the partition col will be removed in the subsequent files

unical1988 replied:
Now there are three errors for sync (in case they sound familiar) namely; for

-  **Hudi**: `replace metadata can only be set once`
- **Iceberg**: `org.apache.iceberg.exceptions.ValidationException: Invalid source type string for transform: year`
- **Delta**:` Cannot change the metadata more than once in a transaction`

unical1988 replied:
In this case, wouldn't it cause Hudi to write the YEAR (yyyy) as values for the timestamp column instead of the actual timestamp data?

the-other-tim-brown replied:
The table formats generally moved away from this approach to partitioning so it cannot be a one to one comparison. Hudi has an option to drop the field with [hoodie.datasource.write.drop.partition.columns](https://hudi.apache.org/docs/configurations/#hoodiedatasourcewritedroppartitioncolumns)

unical1988 replied:
the stated config does not seem to have any effects: the error is still there: T`estParquetConversionSource.testFilePartitionedData:268->checkDatasetEquivalenceWithFilter:286->checkDatasetEquivalence:398 Datasets are not equivalent when reading from Spark. Source: PARQUET, Target: HUDI ==> expected: <[{"id":1,"name":"Alice","age":30,"timestamp":"2025-08-20T01:10:33.725Z"}, {"id":2,"name":"Bob","age":24,"timestamp":"2025-08-20T01:10:34.741Z"}, {"id":3,"name":"Charlie","age":35,"timestamp":"2025-08-20T01:10:35.741Z"}, {"id":4,"name":"David","age":29,"timestamp":"2025-08-20T01:10:36.741Z"}, {"id":5,"name":"Eve","age":22,"timestamp":"2025-08-20T01:10:37.741Z"}]> but was: <[{"id":1,"name":"Alice","age":30,"timestamp":"2025"}, {"id":2,"name":"Bob","age":24,"timestamp":"2025"}, {"id":3,"name":"Charlie","age":35,"timestamp":"2025"}, {"id":4,"name":"David","age":29,"timestamp":"2025"}, {"id":5,"name":"Eve","age":22,"timestamp":"2025"}]>`

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/reflection/ReflectionUtils.java:
```diff
@@ -43,17 +44,31 @@ public static <T> T createInstanceOfClass(String className, Object... constructo
         return clazz.newInstance();
       }
-      Class<?>[] constructorArgTypes =
+      /*Class<?>[] constructorArgTypes =
```
Ok

unical1988 commented:
@the-other-tim-brown do you know this error in Hudi? `Failed to retrieve list of partition from metadata org.apache.hudi.exception.HoodieException: Exception when reading log file` ?

the-other-tim-brown commented on pom.xml:
```diff
@@ -60,15 +60,17 @@
         <!-- To ensure build content is re-producible -->
         <project.build.outputTimestamp>2025-01-01T00:00:00Z</project.build.outputTimestamp>
-        <maven.compiler.target>8</maven.compiler.target>
+        <maven.compiler.source>17</maven.compiler.source><!--should be 8 for CI-->
```
We need to keep the java8 targets

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/delta/DeltaStatsExtractor.java:
```diff
@@ -111,6 +114,10 @@ public String convertStatsToDeltaFormat(
             .maxValues(getMaxValues(validColumnStats))
             .nullCount(getNullCount(validColumnStats))
             .build();
+    // Binary type in Parquet stats serialization is handled with BinarySerializer()
```
Did you validate that this is how Delta is serializing the binary stats?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/HudiConversionTarget.java:
```diff
@@ -373,6 +373,7 @@ private CommitState(
     public void setReplaceMetadata(BaseFileUpdatesExtractor.ReplaceMetadata replaceMetadata) {
       if (!writeStatuses.isEmpty() || !partitionToReplacedFileIds.isEmpty()) {
         throw new IllegalArgumentException("Replace metadata can only be set once");
+        // is of size 1
```
nitpick: remove this

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/PathBasedPartitionValuesExtractor.java:
```diff
@@ -81,40 +77,40 @@ public List<PartitionValue> extractPartitionValues(
       case HOUR:
         return parseDate(
-            remainingPath, pathToPartitionFieldFormat.get(field.getSourceField().getPath()));
+            basePath, field.getSourceField().getPath(), ""); // TODO adjust for other cases
```
Is this TODO completed?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,252 @@
+
+@Builder
+public class ParquetConversionSource implements ConversionSource<Long> {
+  @Builder.Default
```
The builder does not apply to static variables, only instance variables. This annotation should be removed or the variables should be made to instance variables

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,252 @@
+            () -> new IllegalStateException("No file found at " + Long.valueOf(modificationTime)));
+  }
+
+  private Collection<LocatedFileStatus> getParquetFiles(Configuration hadoopConf, String basePath) {
```
Can we update this to return a stream to avoid requiring all files to be loaded into memory at once?

unical1988 replied:
ok, done

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSourceProvider.java:
```diff
@@ -0,0 +1,43 @@
+import org.apache.xtable.conversion.SourceTable;
+import org.apache.xtable.hudi.PathBasedPartitionSpecExtractor;
+
+/** A concrete implementation of {@link ConversionSourceProvider} for Delta Lake table format. */
```
I think the comment here needs to be updated, it is not for Delta Lake.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetPartitionSpecExtractor.java:
```diff
@@ -0,0 +1,74 @@
+@AllArgsConstructor
+public class ParquetPartitionSpecExtractor implements PathBasedPartitionSpecExtractor {
+  private static final ParquetPartitionSpecExtractor INSTANCE =
+      new ParquetPartitionSpecExtractor(new ArrayList<>());
```
How will the list get update for the table that is being processed?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetPartitionSpecExtractor.java:
```diff
@@ -0,0 +1,74 @@
+  @Override
+  public Map<String, String> getPathToPartitionFieldFormat() {
+    Map<String, String> pathToPartitionFieldFormat = new HashMap<>();
+    // ParquetSourceConfig partitionConfig =
+    // ParquetSourceConfig.fromPartitionFieldSpecConfig(partitionFieldSpecConfig);
```
Remove this comment?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -270,12 +267,11 @@ public InternalSchema toInternalSchema(Type schema, String parentPath) {
           String fieldName = parquetField.getName();
           Type.ID fieldId = parquetField.getId();
-          currentRepetition = parquetField.getRepetition();
+          // currentRepetition = parquetField.getRepetition();
```
Why is this change required?

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/ITConversionController.java:
```diff
@@ -483,55 +605,6 @@ public void testTimeTravelQueries(String sourceTableFormat) throws Exception {
-        .collect(Collectors.toList());
-  }
-
-  private static Stream<Arguments> provideArgsForPartitionTesting() {
```
A lot of the changes to this file are just moving methods around. Let's avoid doing this going forward unless it is required or makes the code a lot easier to read. It adds more overhead to the review for me to go line by line through this to check what changes there are to the test setups.

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/ITConversionController.java:
```diff
@@ -188,6 +306,10 @@ private ConversionSourceProvider<?> getConversionSourceProvider(String sourceTab
     }
   }
+  /*
+     test for Parquet file conversion
```
Is there a test case added?

the-other-tim-brown replied:
Let's remove these changes then if they are not required

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/iceberg/IcebergColumnStatsConverter.java:
```diff
@@ -68,12 +72,28 @@ public Metrics toIceberg(Schema schema, long totalRowCount, List<ColumnStat> fie
           if (columnStats.getRange().getMinValue() != null) {
-            lowerBounds.put(
-                fieldId, Conversions.toByteBuffer(fieldType, columnStats.getRange().getMinValue()));
+            if (fieldType.toString() == "string" && format == "APACHE_PARQUET") {
```
The Binary type in Parquet stats only requires this type-casting. That's why I use format to check if the conversion is needed or not

the-other-tim-brown replied:
Is Binary a single byte in this case?
If so, we need to update the spec in the `Range` class to specify how that will be represented. There is no need to do any casting in the other converters since the types are consistent across formats in XTable. That is the purpose of the spec.

unical1988 replied:
the only way for this converter to read the parquet stat (in case of String type) is to do this kind of casting. Now I don't know if this can be done elsewhere e.g., in Range class?

unical1988 replied:
@the-other-tim-brown seems to cause an error in Hudi if I convert in ParquetStatsExtractor.

the-other-tim-brown replied:
If the field is a string, it must be translated to a string when extracting from the source. This is the spec that has worked for 3 other table formats, I think it should work here as well.

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -0,0 +1,252 @@
+    InternalTable table = getMostRecentTable(parquetFiles);
+    return InternalSnapshot.builder()
+        .table(table)
+        .sourceIdentifier(getCommitIdentifier(1L)) // TODO check for version number instead
```
this: `org.apache.parquet.io.api.Binary;`

unical1988 replied:
not sure I understand "last time"?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/BaseFileUpdatesExtractor.java:
```diff
@@ -230,13 +230,14 @@ private WriteStatus toWriteStatus(
 
   private Map<String, HoodieColumnRangeMetadata<Comparable>> convertColStats(
-      String fileName, List<ColumnStat> columnStatMap) {
+      String fileName, List<ColumnStat> columnStatMap, String fileFormat) {
```
We cannot have this be dependent on fileFormat as mentioned before. The intermediate object should be standardized so it is not dependent on source format

unical1988 replied:
It isnt, i forgot to remove the parameter

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -295,15 +298,23 @@ public InternalSchema toInternalSchema(Type schema, String parentPath) {
               .fields(subFields)
-              .isNullable(isNullable(schema.asGroupType()))
+              .isNullable(
+                      isNullable(schema.asGroupType())) // false isNullable(schema.asGroupType()) (TODO causing
```
Right now this is returning that the top level record schema is nullable which should not be the case. Do we need some special handling for this case?

unical1988 replied:
As is, it does not seem to cause any error

unical1988 replied:
It seems indeed that this requires special handling in the case of passed schema (record)

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -112,17 +119,59 @@ public static Map<ColumnDescriptor, List<ColumnStat>> getStatsForFile(ParquetMet
                                     Range.vector(
-                                        columnMetaData.getStatistics().genericGetMin(),
-                                        columnMetaData.getStatistics().genericGetMax()))
+                                        columnMetaData.getPrimitiveType().getPrimitiveTypeName()
+                                                == PrimitiveType.PrimitiveTypeName
+                                                    .BINARY // TODO how about DECIMAL, JSON, BSON
+                                            // and ENUM logicalTypes?
+                                            ? columnMetaData
+                                                        .getPrimitiveType()
+                                                        .getLogicalTypeAnnotation()
+                                                    != null
+                                                ? columnMetaData
+                                                        .getPrimitiveType()
+                                                        .getLogicalTypeAnnotation()
+                                                        .toString()
+                                                        .equals("STRING")
+                                                    ? new String(
+                                                        ((Binary)
+                                                                columnMetaData
+                                                                    .getStatistics()
+                                                                    .genericGetMin())
+                                                            .getBytes(),
+                                                        StandardCharsets.UTF_8)
+                                                    : columnMetaData.getStatistics().genericGetMin()
+                                                : columnMetaData.getStatistics().genericGetMin()
+                                            : columnMetaData
+                                                .getStatistics()
+                                                .genericGetMin(), // if stats are string convert to
+                                        // litteraly a string stat and
```
Let's move this conversion logic into a helper method and then add a new GH Issue for handling any other logical types.

unical1988 replied:
Ok

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -132,19 +181,23 @@ public static InternalDataFile toInternalDataFile(Configuration hadoopConf, Path
+                      parquetMetadataExtractor.readParquetMetadata(hadoopConf, file.getPath()),
+                      file.getPath().toString())),
+              parentPath.toString());
     } catch (java.io.IOException e) {
```
I think this is just hiding the exception currently and we need to fix this.

the-other-tim-brown replied:
It catches the exception but does not throw any new exception. The user will not know of the errors.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/HudiPathUtils.java:
```diff
@@ -28,4 +28,8 @@ public static String getPartitionPath(Path tableBasePath, Path filePath) {
     return endIndex <= startIndex ? "" : pathStr.substring(startIndex, endIndex);
   }
+
+  public static String getPartitionPathValue(Path tableBasePath, Path filePath) {
+    return getPartitionPath(tableBasePath, filePath).split("=")[1];
+  }
```
What if there are multiple `=` in the path? Like for `/some/path/year=2025/month=08/`

unical1988 replied:
Nested partition are not handled by that helper, will need some tweaking

the-other-tim-brown replied:
Also this should be moved to the parquet utils directly since Hudi has handling for multiple partition fields already

unical1988 replied:
Btw, since the sourcefield name (timestamp) is different from the partition col (year) i am not sure Hudi logic extracts the values of partition from the path?

unical1988 replied:
I cannot use sourceField as the year col, since year col is not part of the schema

the-other-tim-brown replied:
We use the user provided partition spec for Hudi to determine the mapping of field in the data to the partition path pattern. For example, a field in the data called `ts` is mapped to `year=YYYY/month=MM/day=DD` with a config `ts:DAY:year=YYYY/month=MM/day=DD`.

unical1988 replied:
I changed the partitioning for parquet while keeping functions from hudi, i think it will be able to deal with nested partitions.

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestSparkParquetTable.java:
```diff
@@ -0,0 +1,117 @@
+
+import org.apache.xtable.GenericTable;
+
+public class TestSparkParquetTable implements GenericTable<Group, String> {
```
Is this being used by the tests?

unical1988 replied:
Yes

unical1988 commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetConversionSource.java:
```diff
@@ -0,0 +1,434 @@
+import org.apache.xtable.hudi.HudiTestUtil;
+import org.apache.xtable.model.sync.SyncMode;
+
+public class TestParquetConversionSource {
```
Yes renaming the test class with prefix IT solved it

unical1988 commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -0,0 +1,435 @@
+
+  @ParameterizedTest
+  @MethodSource("provideArgsForFilePartitionTesting")
+  public void testFilePartitionedData(
```
Ok

unical1988 commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -0,0 +1,435 @@
+              null);
+      ConversionController conversionController =
+          new ConversionController(jsc.hadoopConfiguration());
+      conversionController.sync(conversionConfig, conversionSourceProvider);
```
Not sure how to perform multiple syncs at once?

unical1988 replied:
Suppose I am writing new files for subsequent syncs, should they partitioned as was the case for the first data or not? should the sparksession be the same when writing the parquet for the second sync? how about the folder where to write (same as tempDir?)

unical1988 commented on pom.xml:
```diff
@@ -712,6 +730,16 @@
                     </execution>
                 </executions>
                 <configuration>
+                    <argLine>
```
yes it is

the-other-tim-brown replied:
interesting, why is this required all of a sudden?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetPartitionValueExtractor.java:
```diff
@@ -0,0 +1,115 @@
+      InternalPartitionField field, String remainingPath, int totalNumberOfPartitions, int index) {
+    switch (field.getTransformType()) {
+      case YEAR:
+      case MONTH: // TODO split and get the value of month from pathToPartitionFieldFormat
```
let's add a ticket for all of these TODO items

unical1988 replied:
that was the old partitioning logic, it is the same for all cases (as Hudi's case), just removed the comments here.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -97,6 +98,11 @@ public InternalSchema toInternalSchema(Type schema, String parentPath) {
           // PrimitiveTypes
+        case INT96: // TODO check logicaltypes of INT96
+          metadata.put(
+              InternalSchema.MetadataKey.TIMESTAMP_PRECISION, InternalSchema.MetadataValue.MILLIS);
```
int96 is used with nanosecond precision

unical1988 replied:
done

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -284,6 +290,7 @@ public InternalSchema toInternalSchema(Type schema, String parentPath) {
               == 1) { // TODO Tuple (many subelements in a list)
             newDataType = subFieldSchema.getDataType();
             elementName = subFieldSchema.getName();
+            // subFields = subFieldSchema.getFields();
```
Remove this?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsConverterUtil.java:
```diff
@@ -0,0 +1,69 @@
+    Object returnedObj = null;
+    PrimitiveType primitiveType = columnMetaData.getPrimitiveType();
+    switch (primitiveType.getPrimitiveTypeName()) {
+      case BINARY: // TODO check if other primitiveType' needs to be handled as well
```
bump on this request

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/hudi/TestHudiPartitionValuesExtractor.java:
```diff
@@ -237,9 +237,9 @@ public void testDateAndTimeFormats(
 
     Map<String, String> pathToPartitionFieldFormat = new HashMap<>();
-    pathToPartitionFieldFormat.put(column.getSourceField().getPath(), format);
+    pathToPartitionFieldFormat.put(column.getSourceField().getName(), format);
```
Let's keep this as `path`, it is required for Hudi which can use a nested field for a partition

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -0,0 +1,566 @@
+                // "year=YYYY/month=MM/day=DD"
+                // "timestamp:YEAR:year=YYYY",
+                partitionConfig,
+                levelFilter)));
```
The filter does not line up with the partition config used

unical1988 replied:
filter here is not used, will remove it

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -0,0 +1,566 @@
+      conversionController.sync(conversionConfig, conversionSourceProvider);
+      checkDatasetEquivalenceWithFilter(
+          sourceTableFormat, tableToClose, targetTableFormats, filter);
+    } catch (URISyntaxException e) {
+      throw e;
```
No need to catch this, you can just let it throw

unical1988 replied:
ok

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -0,0 +1,566 @@
+        conversionControllerAppended.sync(conversionConfigAppended, conversionSourceProvider);
+      }
+
+    } catch (URISyntaxException e) {
+      throw e;
```
Similarly here, just let it throw

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -0,0 +1,566 @@
+      } else {
+        assertFalse(dataset1Rows.isEmpty());
+      }
+      if (!format.equals("HUDI")) {
```
If Hudi is not supported, we can just add that as a limitation and exclude it from the sync targets

unical1988 replied:
it is supported, but dataset equivalence are not the same for all formats given the parition approach of Hudi

the-other-tim-brown replied:
Ok, let's make sure we run the query for the non-partitioned dataset then?

unical1988 replied:
yes I am doing that in tests

the-other-tim-brown replied:
Where is that happening? The code is commented out currently

unical1988 replied:
we also talked about this and said the partitioning will not align since the approaches moved away from this old way of doing. therefore checking equivalent over Hudi is disabled (since the partition columns won't be equal).

unical1988 replied:
Done

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/hudi/PathBasedPartitionValuesExtractor.java:
```diff
@@ -88,7 +84,9 @@ private PartialResult parsePartitionPath(
-            remainingPath, pathToPartitionFieldFormat.get(field.getSourceField().getPath()));
+            remainingPath,
+            pathToPartitionFieldFormat.get(
+                field.getSourceField().getPath())); // changed from getPath()
```
yes reverted it back the getPath, will remove the comment

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/iceberg/IcebergPartitionValueConverter.java:
```diff
@@ -190,17 +190,20 @@ public PartitionKey toIceberg(
       return null;
     }
-    Map<String, PartitionValue> nameToPartitionInfo =
+    Map<String, List<PartitionValue>> nameToPartitionInfoList =
```
we discussed this and said that the paritition column is timestamp, if you partition by year and month, the reference to timestamp col will be listed twice

the-other-tim-brown replied:
Yes, this is a bug. Must be finished to merge the PR. I am blocking on this

unical1988 replied:
the test passes, and we agreed on the approach, no need to block

the-other-tim-brown replied:
We did not agree on this approach. I also noted that the test passes but produces the wrong output. The code for the other formats does not change because of the introduction of a new source. I have been saying this from the beginning when you tried to force all targets to be aware of the source format.

unical1988 commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -0,0 +1,566 @@
+            .option("recursiveFileLookup", "true")
+            .option("pathGlobFilter", "*.parquet")
+            .parquet(sourceTable.getDataPath()); // + "/**/*.parquet");
+    // .orderBy(sourceTable.getOrderByColumn())
+    // .filter(filterCondition);
```
Sure

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetPartitionValueExtractor.java:
```diff
@@ -0,0 +1,115 @@
+import org.apache.xtable.model.stat.Range;
+
+/** Partition value extractor for Parquet. */
+public class ParquetPartitionValueExtractor extends PathBasedPartitionValuesExtractor {
```
Is there any tests validating this code? I ran the integration test and see unexpected partition values being set in iceberg.

unical1988 replied:
partition values for iceberg and other formats have been tested over Month and Year as the end-to-end test shows.

the-other-tim-brown replied:
Where is the validation? The filtering on the partition value is skipped, this is how we asserted that the partition values are properly set originally. Another option is to add unit tests to quick show the output is set.

unical1988 replied:
it is a slightly tweaked approach compared to Hudi's, adding unit tests for the partitionValuesExtractor won't be much different from Hudi's.

unical1988 commented:
@the-other-tim-brown I have addressed all of your reviewing comments by now. the CI passes. you could safely merge this PR. if you think the underlying code should be aware of any addition to the set of supported formats you could open an issue for a bug issue later on.

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -295,15 +302,30 @@ public InternalSchema toInternalSchema(Type schema, String parentPath) {
               .dataType(InternalType.RECORD)
               .fields(subFields)
-              .isNullable(isNullable(schema.asGroupType()))
+              .isNullable(isNullable) // false isNullable(schema.asGroupType()) (TODO causing
```
Yes done

the-other-tim-brown commented:
Adds support for https://github.com/apache/incubator-xtable/issues/553
